lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -3437,15 +3437,32 @@ LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
   // Get the result type
   auto resultType = getTypeConverter()->convertType(op.getType());
 
+  int64_t numBroadcastedDims = resultShape.size() - selfType.getRank();
+
+  // Result dimension -1 means not changing the size of that dimension.
+  // Adjust it by assigning its inputShape according to the rank difference
+  // between input and result.
   SmallVector<int64_t> inputShape(
       makeShapeTorchCompatible(selfType.getShape()));
-  // Result dimension -1 means not changing the size of that dimension.
-  // Adjust it by assigning its inputShape.
-  for (auto shape : llvm::enumerate(makeShapeTorchCompatible(inputShape))) {
-    auto index = shape.index();
+  for (auto shape : llvm::enumerate(inputShape)) {
+    auto index = shape.index() + numBroadcastedDims;
     if (resultShape[index] == -1)
       resultShape[index] = shape.value();
   }
+
+  // If there are still unknown dimensions, nothing can be done.
+  if (llvm::any_of(resultShape, [&](auto dim) { return dim == -1; })) {
+    return rewriter.notifyMatchFailure(
+        op, "cannot propagate unknown (-1) dimension "
+            "as it is not presented in the input.");
+  }
+
+  // Add 1 to each broadcasted dimension in the input.
+  // Broadcasted dimensions are the outermost ones.
+  SmallVector<int64_t> broadcastedDims(numBroadcastedDims, 1);
+  inputShape.insert(inputShape.begin(), broadcastedDims.begin(),
+                    broadcastedDims.end());
+
   // Check for identity case i.e, for ex: [a, b, c] -> [a, b, c]. If this is
   // true then we can replace the op result with the input operand directly.
   if (llvm::equal(inputShape, resultShape)) {
```

Comment:
If I do `torch.broadcast_to(torch.zeros(1,1,1), [1, -1, -1, -1])`, would I pass this check here and then go into the replaceAllUsesWith branch? It seems that this would be wrong.

Reply:
Your comment is very pertinent, but it still works properly. In fact, the input IR for this is:
```
%int1 = torch.constant.int 1 loc(#loc1)
    %int-1 = torch.constant.int -1 loc(#loc2)
    %0 = torch.prim.ListConstruct %int1, %int-1, %int-1, %int-1 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int> loc(#loc)
    %1 = torch.prim.ListConstruct %int1, %int1, %int1, %int1 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int> loc(#loc3)
    %2 = torch.aten.view %arg0, %1 : !torch.vtensor<[1,1,1],f32>, !torch.list<int> -> !torch.vtensor<[1,1,1,1],f32> loc(#loc3)
    %3 = torch.aten.broadcast_to %2, %0 : !torch.vtensor<[1,1,1,1],f32>, !torch.list<int> -> !torch.vtensor<[1,1,1,1],f32> loc(#loc3)
    return %3 : !torch.vtensor<[1,1,1,1],f32> loc(#loc)
```
which essentially creates a `ViewOp` as input to `BroadcastToOp`. In any case, I will also check that `!numBroadcastedDims` in the `if` part so that we only enter when indeed there is no broadcasting. Does that cover your concern?

Reply:
Yes, sounds good!

Reply:
And can we have the corresponding end2end test for `torch.broadcast_to(torch.zeros(1,1,1), [1, -1, -1, -1])`?